walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,108 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+
+    // Only works with sorted, non-overlapping time ranges
+    // checkTimes prepares data
+    private Collection<TimeRange> invert(Collection<TimeRange> booked_times, long requestDuration){
+        Collection<TimeRange> possible_times = new ArrayList<TimeRange>();
+        int firstTime = 0;
+        for(TimeRange tr : booked_times){
+            TimeRange free = TimeRange.fromStartEnd(firstTime, tr.start(), false);
+
+            // Ensures a meeting ending before the endpoint of a previous meeting cannot allow
+            // an invalid meeting to be scheduled
+            if(tr.start() + tr.duration() > firstTime)
+                firstTime = tr.start() + tr.duration();
+
+            if(free.duration() >= requestDuration)
+                possible_times.add(free);
+        }
+        TimeRange endOfDay = TimeRange.fromStartEnd(firstTime, 1440, false);
+        if(endOfDay.duration() >= requestDuration){
+            possible_times.add(endOfDay);
+        }
+        
+        return possible_times;
+    }
+
+    // This method doesn't seem efficient. I'd appreciate suggestions
+    // TODO: Refactor for efficiency
```

Comment:
Yeah, you're definitely doing a lot here.  What if you make a separate function, `coalesce()` or `clean()` that takes a random jumble of time ranges and converts them to a nice list of ordered, non-overlapping intervals?  That way, this function can just focus on filtering out the relevant events.

I couple of tips for writing this new function: (1) Java has sorting functions built-in, so you don't have to implement insertion-sort yourself, and (2) if you sort the list of events by start time before checking for overlaps, you can do it with one loop over the list and not have to do any inefficent nested loops.